File: frontend/src/api/taskApi.ts
import type { Page, Task, TaskFilters, TaskInput, TaskStatus } from '../types/task'
import { api } from './client'

const BASE = '/api/tasks'

export const taskApi = {
  list(filters: TaskFilters = {}, signal?: AbortSignal): Promise<Page<Task>> {
    return api.get<Page<Task>>(BASE, { ...filters }, signal)
  },

  get(id: number): Promise<Task> {
    return api.get<Task>(`${BASE}/${id}`)
  },

  create(input: TaskInput): Promise<Task> {
    return api.post<Task>(BASE, input)
  },

  update(id: number, input: TaskInput): Promise<Task> {
    return api.put<Task>(`${BASE}/${id}`, input)
  },

  remove(id: number): Promise<void> {
    return api.delete<void>(`${BASE}/${id}`)
  },

  /**
   * The API has no status-count endpoint, so the funnel reads totalElements from a
   * one-row page per status rather than counting a fully loaded list in the browser.
   */
  async countByStatus(status: TaskStatus, signal?: AbortSignal): Promise<number> {
    const page = await api.get<Page<Task>>(BASE, { status, size: 1 }, signal)
    return page.totalElements
  },
}

File: mobile/lib/api/task_api.dart
import '../models/task.dart';
import 'api_client.dart';

class TaskApi {
  TaskApi(this._client);
  final ApiClient _client;

  /// Filtering, search and pagination run on the server: the app never downloads the whole
  /// list to narrow it locally.
  Future<Page<Task>> list({
    int page = 0,
    int size = 20,
    TaskStatus? status,
    Quadrant? quadrant,
    String? search,
  }) async {
    final json = await _client.request('GET', '/api/tasks', query: {
      'page': '$page',
      'size': '$size',
      if (status != null) 'status': statusWire(status),
      if (quadrant != null) 'quadrant': quadrantWire(quadrant),
      if (search != null && search.isNotEmpty) 'search': search,
    }) as Map<String, dynamic>;

    return Page<Task>(
      content: (json['content'] as List)
          .map((t) => Task.fromJson(t as Map<String, dynamic>))
          .toList(),
      totalElements: json['totalElements'] as int,
      totalPages: json['totalPages'] as int,
      number: json['number'] as int,
    );
  }

  Future<Task> create(TaskInput input) async =>
      Task.fromJson(await _client.request('POST', '/api/tasks', body: input.toJson())
          as Map<String, dynamic>);

  Future<Task> update(int id, TaskInput input) async =>
      Task.fromJson(await _client.request('PUT', '/api/tasks/$id', body: input.toJson())
          as Map<String, dynamic>);

  Future<void> remove(int id) => _client.request('DELETE', '/api/tasks/$id');

  /// The API has no status-count endpoint, so the status chart reads totalElements from a
  /// one-row page per status rather than counting a fully loaded list.
  Future<int> countByStatus(TaskStatus status) async {
    final page = await list(size: 1, status: status);
    return page.totalElements;
  }
}
